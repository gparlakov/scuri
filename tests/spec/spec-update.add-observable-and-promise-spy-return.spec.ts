import { Tree } from '@angular-devkit/schematics';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import {
    collectionPath,
    getTestFile,
    getTestFileContents
} from './common';

describe('spec for a class with a method calling a dependency method', () => {
    it('should add EMPTY for the dep method returning an observable', async () => {
        // arrange
        const depsCallsReturnTypesFile = getTestFile('deps-calls-with-return-types.ts')
        const specFileName = depsCallsReturnTypesFile.replace('.ts', '.spec.ts');
        const tree = Tree.empty();
        tree.create(depsCallsReturnTypesFile, getTestFileContents(depsCallsReturnTypesFile));
        tree.create(specFileName, getTestFileContents(specFileName));
        const runner = new SchematicTestRunner('schematics', collectionPath);

        // act
        const result = await runner
            .runSchematicAsync('spec', { name: depsCallsReturnTypesFile, update: true }, tree)
            .toPromise();

        // assert
        const specFile = result!.readContent(depsCallsReturnTypesFile.replace('.ts', '.spec.ts'));
        expect(specFile).toBeDefined();
        expect(specFile).toMatch('service.observableReturning.and.returnValue(EMPTY)')
        expect(specFile).toMatch('service.promiseReturning.and.returnValue(new Promise(res => {}))');
    });

    it('should add EMPTY for the dep method returning an observable when dependency is already part of the spec', async () => {
        // arrange
        const tree = Tree.empty();
        const fileName = getTestFile('deps-calls-with-return-types-dep-included.ts');
        const specFileName = fileName.replace('.ts', '.spec.ts');
        tree.create(fileName, getTestFileContents(fileName));
        tree.create(specFileName, getTestFileContents(specFileName));
        const runner = new SchematicTestRunner('schematics', collectionPath);

        // act
        const result = await runner
            .runSchematicAsync('spec', { name: fileName, update: true }, tree)
            .toPromise();

        // assert
        const specFile = result!.readContent(fileName.replace('.ts', '.spec.ts'));
        expect(specFile).toBeDefined();
        expect(specFile).toMatch('service.observableReturning.and.returnValue(EMPTY)')
        expect(specFile).toMatch('service.promiseReturning.and.returnValue(new Promise(res => {}))');
    });
});
