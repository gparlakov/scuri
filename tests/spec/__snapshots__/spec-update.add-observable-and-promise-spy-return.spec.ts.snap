// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spec for a class with a method calling a dependency method should add EMPTY/a Promise for the dep method returning an observable/Promise when dependency is already part of the spec 1`] = `
"import { ExampleComponent } from './deps-calls-with-return-types';
import { ServiceWithMethods } from './deps-calls-with-return-types.dependency';
import { EMPTY, Observable } from 'rxjs';
import { autoSpy } from 'autoSpy';

describe('ExampleComponent', () => {
    it('when aMethod is called it should', () => {
        // arrange
        const { build } = setup().default();
        const e = build();
        // act
        e.aMethod();
        // assert
        // expect(e).toEqual
    });
});

function setup() {
  const service = autoSpy(ServiceWithMethods);
  const x: Observable<void> = EMPTY;
  service.observableReturning.and.returnValue(EMPTY);
  service.promiseReturning.and.returnValue(new Promise(res => {}));
  
  const builder = {
    default() {
      return builder;
    },
    build() {
      return new ExampleComponent(service);
    },
    
  withServiceJustAMethodReturn(j: ClassDescription) {
        service.justAMethod.and.returnValue(j);
        return builder;
    },
    withServiceObservableReturningReturn(o: Observable<string>) {
        service.observableReturning.and.returnValue(o);
        return builder;
    },
    withServicePromiseReturningReturn(p: Promise<string>) {
        service.promiseReturning.and.returnValue(p);
        return builder;
    },};

  return builder;
}
"
`;

exports[`spec for a class with a method calling a dependency method should add methods for changing the property when adding it as a dep 1`] = `
"
import { ExampleComponent } from './component';
import { autoSpy } from 'autoSpy';
import { ServiceWithMethods } from './dependency';
import { EMPTY, Observable, ReplaySubject } from 'rxjs';

describe('ExampleComponent', () => {
    it('when aMethod is called it should', () => {
        // arrange
        const { build } = setup().default();
        const e = build();
        // act
        e.aMethod();
        // assert
        // expect(e).toEqual
    });
    it('when anotherMethod is called it should', () => {
        // arrange
        const { build } = setup().default();
        const e = build();
        // act
        e.anotherMethod();
        // assert
        // expect(e).toEqual
    });


});

function setup() {
  const serviceProperty$ = new ReplaySubject<string>(1);
 let resolvePromiseProp: Function;
 let rejectPromiseProp: Function;
 const servicePromiseProp = new Promise((res, rej) => {
     resolvePromiseProp = res;
     rejectPromiseProp = rej;
 });
 const serviceObservable$ = new ReplaySubject<ClassDescription[]>(1);
 const serviceSubject$ = new ReplaySubject<string>(1);
 const service = autoSpy(ServiceWithMethods, { property$: serviceProperty$, promiseProp: servicePromiseProp, observable$: serviceObservable$, subject$: serviceSubject$ });
  service.observableReturning.and.returnValue(EMPTY);
  service.promiseReturning.and.returnValue(new Promise(res => {}));
  const service = autoSpy(ServiceWithMethods);

  const builder = {
    service,
    withJustReturns(j: ClassDescription) {
      // this should still be in for the just-a-method method
      service.justAMethod.and.returnValue(j);
      return builder;
    },
    withPromiseProp() {
      // this should still be in for the promise prop method
      return builder;
    },
    default() {
      return builder;
    },
    build() {
      return new ExampleComponent(service);
    },
    
  withServiceObservableReturningReturn(o: Observable<string>) {
        service.observableReturning.and.returnValue(o);
        return builder;
    },
    withServicePromiseReturningReturn(p: Promise<string>) {
        service.promiseReturning.and.returnValue(p);
        return builder;
    },
    withServiceProperty$(p$: Observable<string>) {
        p$.subscribe({
            next: (v) => serviceProperty$.next(v),
            error: (e) => serviceProperty$.error(e),
            complete: () => serviceProperty$.complete()
        });
        return builder;
    },
    withServicePromiseProp(p: Promise<string>) {
        p
            .then((v) => resolveServicePromiseProp(v))
            .catch((e) => rejectServicePromiseProp(e));
        return builder;
    },
    withServiceObservable$(o$: Observable<ClassDescription[]>) {
        o$.subscribe({
            next: (v) => serviceObservable$.next(v),
            error: (e) => serviceObservable$.error(e),
            complete: () => serviceObservable$.complete()
        });
        return builder;
    },
    withServiceSubject$(s$: Observable<string>) {
        s$.subscribe({
            next: (v) => serviceSubject$.next(v),
            error: (e) => serviceSubject$.error(e),
            complete: () => serviceSubject$.complete()
        });
        return builder;
    },};

  return builder;
}
"
`;

exports[`spec for a class with a method calling a dependency method should add methods for changing the property when adding it as a dep when builder ends with comma 1`] = `
"
import { ExampleComponent } from './component';
import { autoSpy } from 'autoSpy';
import { ServiceWithMethods } from './dependency';
import { EMPTY, Observable, ReplaySubject } from 'rxjs';

describe('ExampleComponent', () => {
    it('when aMethod is called it should', () => {
        // arrange
        const { build } = setup().default();
        const e = build();
        // act
        e.aMethod();
        // assert
        // expect(e).toEqual
    });
    it('when anotherMethod is called it should', () => {
        // arrange
        const { build } = setup().default();
        const e = build();
        // act
        e.anotherMethod();
        // assert
        // expect(e).toEqual
    });


});

function setup() {
  const serviceProperty$ = new ReplaySubject<string>(1);
 let resolvePromiseProp: Function;
 let rejectPromiseProp: Function;
 const servicePromiseProp = new Promise((res, rej) => {
     resolvePromiseProp = res;
     rejectPromiseProp = rej;
 });
 const serviceObservable$ = new ReplaySubject<ClassDescription[]>(1);
 const serviceSubject$ = new ReplaySubject<string>(1);
 const service = autoSpy(ServiceWithMethods, { property$: serviceProperty$, promiseProp: servicePromiseProp, observable$: serviceObservable$, subject$: serviceSubject$ });
  service.observableReturning.and.returnValue(EMPTY);
  service.promiseReturning.and.returnValue(new Promise(res => {}));
  const service = autoSpy(ServiceWithMethods);

  const builder = {
    service,
    withJustReturns(j: ClassDescription) {
      // this should still be in for the just-a-method method
      service.justAMethod.and.returnValue(j);
      return builder;
    },
    withPromiseProp() {
      // this should still be in for the promise prop method
      return builder;
    },
    default() {
      return builder;
    },
    build() {
      return new ExampleComponent(service);
    },
  withServiceObservableReturningReturn(o: Observable<string>) {
        service.observableReturning.and.returnValue(o);
        return builder;
    },
    withServicePromiseReturningReturn(p: Promise<string>) {
        service.promiseReturning.and.returnValue(p);
        return builder;
    },
    withServiceProperty$(p$: Observable<string>) {
        p$.subscribe({
            next: (v) => serviceProperty$.next(v),
            error: (e) => serviceProperty$.error(e),
            complete: () => serviceProperty$.complete()
        });
        return builder;
    },
    withServicePromiseProp(p: Promise<string>) {
        p
            .then((v) => resolveServicePromiseProp(v))
            .catch((e) => rejectServicePromiseProp(e));
        return builder;
    },
    withServiceObservable$(o$: Observable<ClassDescription[]>) {
        o$.subscribe({
            next: (v) => serviceObservable$.next(v),
            error: (e) => serviceObservable$.error(e),
            complete: () => serviceObservable$.complete()
        });
        return builder;
    },
    withServiceSubject$(s$: Observable<string>) {
        s$.subscribe({
            next: (v) => serviceSubject$.next(v),
            error: (e) => serviceSubject$.error(e),
            complete: () => serviceSubject$.complete()
        });
        return builder;
    },};

  return builder;
}
"
`;
